book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+
+### for forward compatibility
+
+We can make sure mixed versions of cluster doesn't break for given change in
+this way:
+
+1. Only backport the newer test fixtures without application code changes
+2. Run deserialize test with fixture of REV+1, REV+2, ... REV+LATEST
+3. Run serialize test with fixture of REV+1, REV+2, ... REV+LATEST
+
+
+# Developer's work flow
+
+We will end up accumulating increasing number of technical debt over time to
+support older ABI item revisions. As a general rule of thumb for implementation
+design, we always normalize the older serialized data into the runtime/internal
+representation which closely reflects the latest revision of given ABI item as
+much as possible with additional conversion logic.
+
+Because it's more expensive to maintain this ABI test than normal one, normal
+unit test is preferred for general unit testing as before. ABI unit test should
+be restricted to the bare minimal set of variation of serialization.
+
+## Change an existing compatibility item:
+
+1. Change `#[solana_abi]` `struct`s as usually
+2. Depending on its maintenance tier, decide hard-fork or normal update
+3. (For hard-fork,) decide scheduled release time based on epochs
+4. Write the code to differentiate depending on
+   [version](#how to detect versions)
+5. Generate new test fixture for the new compatibility item revision and check
+   the diff.
+
+## Create new compatibility item:
+
+1. As there is no free standing system component, this thing is actually
+   related to the other part of system in some way. Consider what compatibility
+   status and maintenance tier is expected for the relation regarding adding
+   something new.
+2. Write entirely new struct/module code with `#[solana_abi]`
+3. Write entirely new serialization/deserialization unit test code
+4. Generate fixture for the first revision and check the diff.
+
+## How to generate test fixtures for newer versions
+
+run `cargo test` with specific environment like `SOLANA_UPDATE_ABI_FIXTURES=1`.
+Then, these fixtures will be automatically updated and review with `git diff`
+
+# Implemention remarks
+
+We'll end up heavily relying on macro magics. For precedent, `ink' from the
+Parity Technologies [2], could be informational.
+
+# Considerations
+ 
+- how to detect vesions?
+  - Introduce version header? (2 byte for everything)
+  - Implicit version header by optimistic signature verification with version as
+    salt? This detects the version by first verifying signature for blob with
+    (salt=latest revision) and if failed try with (salt=latest revision-1) and
+    so on.
+- how to calculate type digest?
+  - sha1 tuple of (field names, type names), don't consider non derived
+    (De)Serializer; tolerate for incompleteness as mentioned
+- how to integrate well with existing serialization libraries (ie. serde,
+  bincode)?
+- We currently only support amd64, there might be unforseen ABI problems for
```

Comment:
Yes, great point.   Other architectures too (an ARM port for example).   I'm ok with just saying this is out of scope for the purposes of the current design

Reply:
Yeah! I changed to mention this lightly at the start: https://github.com/solana-labs/solana/pull/7524/files#diff-7bbc1dbd50f4144bb25bfb9d3c28257aR42